add a doc page about dynamic forward proxy (#6100)
solo-io/gloo: docs/content/guides/traffic_management/listener_configuration/http_connection_manager/dfp/_index.md

---
title: Dynamic Forward Proxy
weight: 10
---

> This document introduces the **Dynamic Forward Proxy** HTTP filter in Gloo Edge.

Reviewer: Maybe add link for more information?
```suggestion
You can set up an [HTTP Dynamic Forward Proxy (DFP) filter](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_proxy) in Gloo Edge.
```

In a highly dynamic environment with services coming up and down and with no service registry being able to list the available endpoints, one option is to somehow "blindly" route the client requests upstream. 

Another popular use case is to deploy a forward proxy for all egress traffic. This way, you can observe and control outbound traffic. A common security policy applied here is rate-limiting and, of course, gathering access logs.

So there are two sorts of usage for a forward proxy:

> - routing all the traffic of your local network to an exit node to monitor and control all the egress traffic

Reviewer: ```suggestion
- Route all the traffic of your local network to an exit node with the forward proxy, to monitor and control all the egress traffic.
```

> - or selectively apply it on certain routes, which don’t have a pre-defined destination, and dynamically build the final Host value

Reviewer: ```suggestion
- Apply the forward proxy filter only to certain routes, which don’t have a pre-defined destination, and dynamically build the final Host value.
```

> There are a few downsides to such flexibility:

Reviewer: ```suggestion
Before implementing a dynamic forward proxy, consider the downsides to such flexibility:
```

> - since there is no pre-defined {{< protobuf name="gloo.solo.io.Upstream" display="Upstream" >}} to designate the upstream service, you cannot configure failover policies or client load-balancing

Reviewer: ```suggestion
- Because no pre-defined {{< protobuf name="gloo.solo.io.Upstream" display="Upstream" >}} designates the upstream service, you cannot configure failover policies or client load-balancing.
```

> - DNS resolution is done at runtime. Typically, when a domain name is met for the first time, Envoy will pause the request and synchronously resolve this domain to get the endpoints (IP addresses). Then, these entries are put into a local cache

Reviewer: ```suggestion
- DNS resolution is done at runtime. Typically, when Envoy encounters a domain name the first time, Envoy pauses the request and synchronously resolves this domain to get the endpoints (IP addresses). Then, these entries are put into a local cache.
```

> Of course, there are also good reasons why this still makes sense in an API Gateway:

Reviewer: ```suggestion
Of course, you might still decide to use a dynamic forward proxy in an API Gateway for benefits such as the following:
```

> - you will easily get metrics on the traffic going through the proxy

Reviewer: ```suggestion
- You easily get metrics on the egress traffic that goes through the forward proxy.
```

> - you can enforce authentication and authorization policies

Reviewer: ```suggestion
- You can enforce authentication and authorization policies.
```

> - you can leverage other policies available in Gloo Edge Enterprise, like the WAF (Web Application Firewall) or DLP (Data Loss Prevention)

Reviewer: ```suggestion
- You can leverage other policies available in Gloo Edge Enterprise, like Web Application Firewall (WAF) or Data Loss Prevention (DLP).
```

## Enabling the Dynamic Forward Proxy

> First, you need to enable the DFP filter at the Gateway level:

Reviewer: ```suggestion
First, enable the DFP filter in your Gateway configuration.
```

```bash
kubectl -n gloo-system patch gw/gateway-proxy --type merge -p "
spec:
  httpGateway:
    options:
      dynamicForwardProxy: {}
"
```

> Then you need to capture the actual destination of the client request. It can simply be the `Host` header in the most basic setup, but it can be hidden in other client request headers or body parts. In this latter case, you can create that header dynamically, using a transformation template.

Reviewer: ```suggestion
Then, set the actual destination of the client request. The destination can be the `Host` header in the most basic setup. However, the destination might also be hidden in other client request headers or body parts. In this latter case, you can create the header dynamically by using a transformation template.
```

> Below is a simple example showing how you can apply the dynamic forward option to a route:

Reviewer: ```suggestion
Review the following basic example to see how you can apply the dynamic forward option to a route.
```
